sourcemap/fleet_store.go

```diff
@@ -71,35 +79,74 @@ func newFleetStore(
 	fleetCfg *config.Fleet,
 	cfgs []config.SourceMapMetadata,
 ) (fleetStore, error) {
-	// TODO(stn): Add support for multiple fleet hosts
-	// cf. https://github.com/elastic/apm-server/issues/5514
-	host := fleetCfg.Hosts[0]
-	fleetURLs := make(map[key]string)
+	sourceMapURLs := make(map[key]string)
+	fleetBaseURLs := make([]string, len(fleetCfg.Hosts))
 
 	for _, cfg := range cfgs {
 		k := key{cfg.ServiceName, cfg.ServiceVersion, cfg.BundleFilepath}
-		u, err := common.MakeURL(fleetCfg.Protocol, cfg.SourceMapURL, host, 8220)
+		sourceMapURLs[k] = cfg.SourceMapURL
+	}
+
+	for i, host := range fleetCfg.Hosts {
+		baseURL, err := common.MakeURL(fleetCfg.Protocol, "", host, defaultFleetPort)
 		if err != nil {
 			return fleetStore{}, err
 		}
-		fleetURLs[k] = u
+		fleetBaseURLs[i] = baseURL
 	}
+
 	return fleetStore{
-		apikey:    "ApiKey " + fleetCfg.AccessAPIKey,
-		fleetURLs: fleetURLs,
-		c:         c,
+		apikey:        "ApiKey " + fleetCfg.AccessAPIKey,
+		fleetBaseURLs: fleetBaseURLs,
+		sourceMapURLs: sourceMapURLs,
+		c:             c,
+		rng:           rand.New(rand.NewSource(time.Now().UnixNano())),
 	}, nil
 }
 
 func (f fleetStore) fetch(ctx context.Context, name, version, path string) (string, error) {
 	k := key{name, version, path}
-	fleetURL, ok := f.fleetURLs[k]
+	sourceMapURL, ok := f.sourceMapURLs[k]
 	if !ok {
 		return "", fmt.Errorf("unable to find sourcemap.url for service.name=%s service.version=%s bundle.path=%s",
 			name, version, path,
 		)
 	}
 
+	idxs := f.rng.Perm(len(f.fleetBaseURLs))
+	respReceived := make(chan string, 1)
+	fetchErrReturned := make(chan error, 1)
+
+	for idx := range idxs {
```

Comment:
Since we're making requests to all the URLs in parallel, is there a reason to use a random number generator?

Reply:
I think it's still necessary to use a random number generator: though the go routines are executed concurrently, they are started serially. Over time, I would bet that the tiny difference in start times would become significant and we'd see the go routines started for hosts towards the beginning of the slice finish before those towards the end. I need to research go routines more, but if they are terminated when the surrounding function returns, some fleet hosts with higher slice indexes might never receive a request.
I've also read that go programs don't necessarily use multiple os threads and that the `GOMAXPROCS` default value is 1. So if this were running using a single os thread, the go routines wouldn't be running in parallel and only the fleet hosts with lower index values (or perhaps just the first one) would serve a request.
Let me know if I'm mistaken, I'm still just a go beginner 😃

Reply:
>though the go routines are executed concurrently, they are started serially. Over time, I would bet that the tiny difference in start times would become significant and we'd see the go routines started for hosts towards the beginning of the slice finish before those towards the end.

This could be true! In general, I would assume that the time to start a go routine is a lot lower than the time to finish a network request. I can't find any numbers atm for how long it takes to spawn a go routine, but would assume a network request to take somewhere on the order of milliseconds, and spawning a go routine is probably somewhere in the microseconds realm.

>I need to research go routines more, but if they are terminated when the surrounding function returns, some fleet hosts with higher slice indexes might never receive a request.

Without coordination, go routines exist outside of the lifecycle of the function that starts them (except for the main goroutine, goroutines have no parents or children, in contrast to an os thread or process). In this case, each index will create a goroutine and execute a network request. Since the code below, which reads the result, is exiting as soon as the first response is received (successful or not), the next network request will send to the buffered channel, and any remaining go routines will be stuck forever attempting to send on the full channel.

>I've also read that go programs don't necessarily use multiple os threads and that the GOMAXPROCS default value is 1. So if this were running using a single os thread, the go routines wouldn't be running in parallel and only the fleet hosts with lower index values (or perhaps just the first one) would serve a request.

There may be a single OS thread that performs CPU bound work, but for network i/o the requests will all be started and then the go runtime will most likely context switch to a different, non-waiting go routine (there are a few well-defined places where the go runtime will context switch). Depending on which web request completes first, that go routine will be "woken up" and continue processing.

The default value for `GOMAXPROCS` is the number of CPUs on the system: https://pkg.go.dev/runtime#GOMAXPROCS